lib/charms/traefik_k8s/v1/ingress_per_unit.py

```diff
@@ -679,6 +707,11 @@ def __init__(
         )
 
     def _handle_relation(self, event: RelationEvent):
+        rel = event.relation
+        if not isinstance(rel, Relation):  # type guard
+            raise RuntimeError('event.relation is not Relation but {}'.format(rel))
+        if not _relation_data_exists(rel):
+            return event.defer()
```

Comment:
Do we need to defer, or could we just return? A `relation-created` or `relation-joined` which hasn't set any data yet won't have any when the deferred event is re-scheduled, so it will just re-fire and  cycle here.

When data is actually set, `relation-changed` will hit here and actually have data, so need to defer.

Reply:
my thought was:
- relation-joined, no data -> defer
- [other events], still no data -> re-defer
- [relation-changed], now we do have data -> finally consume the deferred relation-joined event.
But, I reckon, we could just as well wait until we get the relation-changed event to signal that there is data.
In that case, what is the point of listening to -joined and -created at all, when we could just wait for -changed?